examples/flax/text-classification/run_flax_glue.py

```diff
@@ -493,58 +506,88 @@ def eval_step(state, batch):
     # make sure weights are replicated on each device
     state = replicate(state)
 
-    for epoch in range(1, num_epochs + 1):
-        logger.info(f"Epoch {epoch}")
-        logger.info("  Training...")
+    step_per_epoch = len(train_dataset) // train_batch_size
+    total_steps = step_per_epoch * num_epochs
+    epochs = tqdm(range(num_epochs), desc=f"Epoch ... (0/{num_epochs})", position=0)
+    for epoch in epochs:
 
         train_start = time.time()
         train_metrics = []
+
+        # Create sampling rng
         rng, input_rng = jax.random.split(rng)
 
         # train
-        for batch in glue_train_data_collator(input_rng, train_dataset, train_batch_size):
-            state, metrics, dropout_rngs = p_train_step(state, batch, dropout_rngs)
-            train_metrics.append(metrics)
-        train_time += time.time() - train_start
-        logger.info(f"    Done! Training metrics: {unreplicate(metrics)}")
-
-        logger.info("  Evaluating...")
-
-        # evaluate
-        for batch in glue_eval_data_collator(eval_dataset, eval_batch_size):
-            labels = batch.pop("labels")
-            predictions = p_eval_step(state, batch)
-            metric.add_batch(predictions=chain(*predictions), references=chain(*labels))
-
-        # evaluate also on leftover examples (not divisible by batch_size)
-        num_leftover_samples = len(eval_dataset) % eval_batch_size
-
-        # make sure leftover batch is evaluated on one device
-        if num_leftover_samples > 0 and jax.process_index() == 0:
-            # take leftover samples
-            batch = eval_dataset[-num_leftover_samples:]
-            batch = {k: jnp.array(v) for k, v in batch.items()}
-
-            labels = batch.pop("labels")
-            predictions = eval_step(unreplicate(state), batch)
-            metric.add_batch(predictions=predictions, references=labels)
-
-        eval_metric = metric.compute()
-        logger.info(f"    Done! Eval metrics: {eval_metric}")
-
-        cur_step = epoch * (len(train_dataset) // train_batch_size)
-
-        # Save metrics
-        if has_tensorboard and jax.process_index() == 0:
-            write_metric(train_metrics, eval_metric, train_time, cur_step)
-
-        # save checkpoint after each epoch and push checkpoint to the hub
-        if jax.process_index() == 0:
-            params = jax.device_get(jax.tree_map(lambda x: x[0], state.params))
-            model.save_pretrained(args.output_dir, params=params)
-            tokenizer.save_pretrained(args.output_dir)
-            if args.push_to_hub:
-                repo.push_to_hub(commit_message=f"Saving weights and logs of epoch {epoch}", blocking=False)
+        for step, batch in enumerate(
+            tqdm(
+                glue_train_data_collator(input_rng, train_dataset, train_batch_size),
+                total=step_per_epoch,
+                desc="Training...",
+                position=1,
+            ),
+            1,
+        ):
```

Comment:
(nit) This feels a bit hard to read. Could we maybe refactor it like this


```python
train_loader = glue_train_data_collator(input_rng, train_dataset, train_batch_size)
for step, batch in enumerate(train_loader, total=step_per_epoch, desc="Training...", position=1), 1):
    ....
```

Reply:
done